### PR TITLE
Reimplement web audio farbling for performance

### PR DESCRIPTION
**Pull request by pilgrim-brave**
Resolves https://github.com/brave/brave-browser/issues/26562

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: chromium_src/third_party/blink/renderer/modules/webaudio/audio_buffer.cc [hunk 1]
```diff
@@ -13,40 +13,30 @@
...
             brave::GetContentSettingsClientFor(context)) {                \
-      if (brave::OptionalAudioFarblingCallback audio_farbling_callback =  \
-              brave::BraveSessionCache::From(*context)                    \
-                  .GetAudioFarblingCallback(settings)) {                  \
-        for (unsigned i = 0; i < count; i++) {                            \
-          dst[i] = audio_farbling_callback->Run(dst[i], i);               \
-        }                                                                 \
+      DOMFloat32Array* destination_array = array.Get();                   \
+      size_t len = destination_array->length();                           \
+      if (len > 0) {                                                      \
```
Comment (goodov) [inline chromium_src/third_party/blink/renderer/modules/webaudio/audio_buffer.cc:23]: this check can be prioritised, i.e. don't even access `GetContentSettingsClientFor` if `len == 0`.

File: chromium_src/third_party/blink/renderer/modules/webaudio/audio_buffer.cc [hunk 2]
```diff
@@ -13,40 +13,30 @@
...
+        brave::BraveSessionCache::From(*context).FarbleAudioChannel(      \
+            settings, destination, len);                                  \
       }                                                                   \
     }                                                                     \
   }
 
+#define BRAVE_AUDIOBUFFER_COPYFROMCHANNEL                                      \
+  if (ExecutionContext* context = ExecutionContext::From(script_state)) {      \
+    if (WebContentSettingsClient* settings =                                   \
+            brave::GetContentSettingsClientFor(context)) {                     \
```
Comment (goodov) [inline chromium_src/third_party/blink/renderer/modules/webaudio/audio_buffer.cc:34]: pass `ExecutionContext::From(script_state)` directly

File: third_party/blink/renderer/core/farbling/brave_session_cache.cc [hunk 3]
```diff
@@ -234,8 +403,46 @@ void BraveSessionCache::Init() {
...
+    return;
+  DCHECK(settings);
+  switch (settings->GetBraveFarblingLevel()) {
+    case BraveFarblingLevel::OFF: {
+      break;
+    }
+    case BraveFarblingLevel::BALANCED: {
+      const uint64_t* fudge = reinterpret_cast<const uint64_t*>(domain_key_);
+      double fudge_factor = 0.99 + ((*fudge / maxUInt64AsDouble) / 100);
+      VLOG(1) << "audio fudge factor (based on session token) = "
```
Comment (goodov) [inline third_party/blink/renderer/core/farbling/brave_session_cache.cc:420]: `DVLOG` or just move this to the constructor. runtime-enabled logs on a hot path are not free.

Comment (pilgrim-brave) [reply third_party/blink/renderer/core/farbling/brave_session_cache.cc:420]: done

File: chromium_src/third_party/blink/renderer/modules/webaudio/audio_buffer.cc (hunk 2 above)
Comment (pilgrim-brave) [reply chromium_src/third_party/blink/renderer/modules/webaudio/audio_buffer.cc:34]: done

File: chromium_src/third_party/blink/renderer/modules/webaudio/audio_buffer.cc (hunk 1 above)
Comment (pilgrim-brave) [reply chromium_src/third_party/blink/renderer/modules/webaudio/audio_buffer.cc:23]: done

File: chromium_src/third_party/blink/renderer/modules/webaudio/audio_buffer.cc [hunk 4]
```diff
@@ -13,40 +13,30 @@
...
-      }                                                                       \
-    }                                                                         \
-  }
-
-#define BRAVE_AUDIOBUFFER_COPYFROMCHANNEL                                 \
+#define BRAVE_AUDIOBUFFER_GETCHANNELDATA                                  \
+  NotShared<DOMFloat32Array> array = getChannelData(channel_index);       \
   if (ExecutionContext* context = ExecutionContext::From(script_state)) { \
     if (WebContentSettingsClient* settings =                              \
             brave::GetContentSettingsClientFor(context)) {                \
```
[the remark this answers is not in the document]
Comment (pilgrim-brave) [reply chromium_src/third_party/blink/renderer/modules/webaudio/audio_buffer.cc:20]: done

File: chromium_src/third_party/blink/renderer/modules/webaudio/analyser_handler.cc [hunk 5]
```diff
@@ -10,14 +10,17 @@
...
-#define BRAVE_ANALYSERHANDLER_CONSTRUCTOR                                    \
-  if (ExecutionContext* context = node.GetExecutionContext()) {              \
-    if (WebContentSettingsClient* settings =                                 \
-            brave::GetContentSettingsClientFor(context)) {                   \
-      analyser_.audio_farbling_callback_ =                                   \
-          brave::BraveSessionCache::From(*context).GetAudioFarblingCallback( \
-              settings);                                                     \
-    }                                                                        \
+#define BRAVE_ANALYSERHANDLER_CONSTRUCTOR                            \
+  if (ExecutionContext* context = node.GetExecutionContext()) {      \
```
[the remark this answers is not in the document]
Comment (pilgrim-brave) [reply chromium_src/third_party/blink/renderer/modules/webaudio/analyser_handler.cc:14]: done

File: chromium_src/third_party/blink/renderer/modules/webaudio/audio_buffer.cc (hunk 2 above)
Comment (goodov) [reply chromium_src/third_party/blink/renderer/modules/webaudio/audio_buffer.cc:36]: I was wrong here. Context is actually used later, I've restored it in these macroses.

File: third_party/blink/renderer/platform/brave_audio_farbling_helper.cc [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+  } else {
+    for (size_t i = 0; i < count; i++) {
+      dst[i] = dst[i] * fudge_factor_;
+    }
+  }
+}
+
+// Calculate values for RealtimeAnalyser::GetFloatTimeDomainData
+void BraveAudioFarblingHelper::FarbleFloatTimeDomainData(
+    float* input_buffer,
```
Comment (iefremov) [inline third_party/blink/renderer/platform/brave_audio_farbling_helper.cc:47]: pls mark all input buffers as `const`

File: third_party/blink/renderer/platform/brave_audio_farbling_helper.cc [hunk 7]
```diff
@@ -126,8 +126,8 @@ void BraveAudioFarblingHelper::FarbleConvertToByteData(
     const float* source,
     unsigned char* destination,
     size_t len,
-    const double min_decibels,
-    const double range_scale_factor) const {
+    double min_decibels,
+    double range_scale_factor) const {
```
Comment (iefremov) [inline third_party/blink/renderer/platform/brave_audio_farbling_helper.cc:130]: also in the header file!

File: third_party/blink/renderer/core/farbling/brave_session_cache.h [hunk 8]
```diff
@@ -90,12 +89,18 @@ class CORE_EXPORT BraveSessionCache final
   bool AllowFontFamily(blink::WebContentSettingsClient* settings,
                        const AtomicString& family_name);
   FarblingPRNG MakePseudoRandomGenerator(FarbleKey key = FarbleKey::kNone);
+  absl::optional<blink::BraveAudioFarblingHelper> GetAudioFarblingHelper(
```
Comment (atuchin-m) [inline third_party/blink/renderer/core/farbling/brave_session_cache.h:92]: Do we need both `GetAudioFarblingHelper` and `UpdateAndGetAudioFarblingHelper`. In fact they are the same now.

**Review by atuchin-m — APPROVED** [review]
LGTM in terms of perf.
In general it's also looks much better than before.

**Review by goodov — APPROVED** [review]

**Pull request merged**